### PR TITLE
replace usages of TestSpanExporter with InMemorySpanExporter

### PR DESCRIPTION
**Pull request by sreeo**
Fixes #
Replaces usages of TestSpanExporter with InMemorySpanExporter

## Changes

Please provide a brief description of the changes here.

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [ ] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry-sdk/src/trace/span_processor.rs [hunk 1]
```diff
@@ -712,16 +713,18 @@ mod tests {
...
-        let exporter = TestSpanExporter::new();
+        let exporter = InMemorySpanExporterBuilder::new().build();
         let mut processor = SimpleSpanProcessor::new(Box::new(exporter.clone()));
-        assert!(!exporter.is_shutdown_called());
+        let span_data = new_test_export_span_data();
+        processor.on_end(span_data.clone());
+        assert_eq!(exporter.get_finished_spans().unwrap()[0], span_data);
         let _result = processor.shutdown();
-        assert!(exporter.is_shutdown_called());
+        assert!(exporter.get_finished_spans().unwrap().is_empty());
```
Comment (cijothomas) [inline opentelemetry-sdk/src/trace/span_processor.rs:727]: nit: Add a comment saying, we indirectly test shutdown is called, by asserting that spans are empty in inmemoryexporter.

File: opentelemetry-sdk/src/trace/span_processor.rs [hunk 2]
```diff
@@ -712,16 +713,18 @@ mod tests {
...
 
     #[test]
     fn simple_span_processor_shutdown_calls_shutdown() {
-        let exporter = TestSpanExporter::new();
+        let exporter = InMemorySpanExporterBuilder::new().build();
         let mut processor = SimpleSpanProcessor::new(Box::new(exporter.clone()));
-        assert!(!exporter.is_shutdown_called());
+        let span_data = new_test_export_span_data();
+        processor.on_end(span_data.clone());
+        assert_eq!(exporter.get_finished_spans().unwrap()[0], span_data);
```
Comment (cijothomas) [inline opentelemetry-sdk/src/trace/span_processor.rs:725]: nit: we only need to validate that finished spans is not empty before shutdown, and is empty after shutdown, to keep the test very focused on its purpose.

File: opentelemetry-sdk/src/trace/span_processor.rs [hunk 3]
```diff
@@ -820,7 +823,10 @@ mod tests {
             (OTEL_BSP_EXPORT_TIMEOUT, Some("2046")),
         ];
         temp_env::with_vars(env_vars.clone(), || {
-            let builder = BatchSpanProcessor::builder(TestSpanExporter::new(), runtime::Tokio);
```
Comment (cijothomas) [inline opentelemetry-sdk/src/trace/span_processor.rs:823]: Are there more places `TestSpanExporter` is used? If none, could we remove that struct itself?

File: opentelemetry-sdk/CHANGELOG.md [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Changelog
 
 ## vNext
-
+- Replace TestExporter with InMemoryExporter [#1624](https://github.com/open-telemetry/opentelemetry-rust/pull/1624)
```
Comment (cijothomas) [inline opentelemetry-sdk/CHANGELOG.md:4]: please remove this, we dont need changelog for internal testing changes. only for end users facing changes, we need to make entry in here. Thanks.

**Review by cijothomas — APPROVED** [review]
Looks good.
Left couple of comments, to be addressed before merge, esp. removing the changelog entry

File: opentelemetry-sdk/src/trace/span_processor.rs (hunk 3 above)
Comment (sreeo) [reply opentelemetry-sdk/src/trace/span_processor.rs:823]: Only used here, removed the struct.

**Review by cijothomas — APPROVED** [review]

**Review by TommyCpp — APPROVED** [review]
fix #1616

**Comment by TommyCpp** [comment]
looks good. Thanks!
